Add FileHandle::truncate and ftruncate

kjbracey opened the pull request:
### Description

Add support for file truncation (or extension) to the abstract API.

Unit tests added.

Implementation added for:

- `File::truncate`
- `FileSystem::file_truncate`
- `FATFileSystem::file_truncate`
- `LittleFileSystem::file_truncate`

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

kjbracey committed:
Add FileHandle::truncate and ftruncate

Add support for file truncation (or extension) to the abstract API.

No hooks to actual implementations in this commit.

kjbracey commented:
Relevant to https://github.com/ARMmbed/mbed-os/issues/7091 and https://github.com/ARMmbed/mbed-os/issues/7010

dannybenor commented:
I am not sure we want the side effects of this feature on flash based file systems

kjbracey commented:
> I am not sure we want the side effects of this feature on flash based file systems

If a particular filing system cannot or does not want to implement the call, then that's fine. Specialised filing systems can have basically arbitrary limitations.

Only fully general-purpose random-access filing systems would be expected to implement every call and usage pattern and thus be usable by every application.  If you are one of those, then truncating is trivial, and extension can be no worse than the application writing all the zeroes manually, which you must support. If you're not general purpose, maybe you don't support changing file lengths after creation at all. That would be fine.

In a specialised system, one potential use of this is for efficient buffer pre-allocation. Eg a system that supports the sequence "open(), write(), write(), write() close()" to create a file might have a fixed-size buffer pre-allocation and thus limit on the total.

But you could have a contract such that if "truncate" was issued before the first write, you set the write buffer to that size - can be used as an API to specify the total amount you're intending to write.

cmonr commented:
@ARMmbed/mbed-os-storage Would y'all mind posting a response to @kjbracey-arm explanation/rationale ?

geky commented:
Sorry about delay, this looked reasonable to me, but I want to toss in the implementations that came out of the discussions with @OscarGarciaF:
- littlefs - https://github.com/ARMmbed/mbed-os/issues/7091#issuecomment-394422272
- fatfs - https://github.com/ARMmbed/mbed-os/issues/7010#issuecomment-392107503

Next chance I get I can bring them in if no one beats me to it (on top of this pr hopefully).

davidsaada commented on platform/FileHandle.h:
```diff
@@ -137,6 +137,17 @@ class FileHandle : private NonCopyable<FileHandle> {
+     * not changed. If the file is extended, the extended area appears as if
+     * it were zero-filled.
+     */
+    virtual int truncate(off_t length)
```
Needs a Doxy description of the length parameter.

davidsaada reviewed (changes requested):
Other than the minor Doxy issue, looks good to me.

dannybenor commented:
We need to have a test for this new feature

dannybenor commented:
Please do not merge

kjbracey committed:
Add ftruncate / fstat(st_size) unit test

kjbracey commented:
Added unit test covering the FileHandle/retarget bits being added here.

Actual functional tests would be a matter for any invidual filing systems implementing it at the minute - we've no general filing system-agnostic test suite afaik.

geky committed:
Added filesystem implementations of truncate

- File::truncate
- FileSystem::file_truncate
- FATFileSystem::file_truncate
- LittleFileSystem::file_truncate

geky commented:
Hey @kjbracey-arm, I tacked on the filesystem system implementations of truncate. Feel free to remove if they're just getting the in way.

Thanks for writing up the documentation and other glue layers.

dannybenor commented:
My general approach is that I do not want to add new functionality to mbedos that is not clearly required. In addition, we need good testing at the file system level.
Will not merge until both conditions are achieved: We know why we need it and we have tests

kjbracey commented:
For this particular call, I find it hard to think of it as new functionality - I think this just covers an odd omission in the existing "support general-purpose filesystems" functionality. My first thought on seeing a request to add it was "whoops, why wasn't that already there?". If we can extend files, and seek, we should be able to truncate them.

`ftruncate` is a pretty core part of the general POSIX filing system API we try to follow, and is already implemented in the two main general-purpose filing systems in the system - it's just not been exposed through the generic mbed OS C++ API.

We've had a couple of user requests to access this for the general-purposes filing systems, and it also has potential utility as a pre-write size hint for key/value-type filing systems as described above.

Agree that it should be added to FAT and LFS tests when we hook it up to them, so @geky's added commit now requires that.

cmonr commented:
@kjbracey-arm @davidsaada @geky @dannybenor 
Considering that this PR has sat for over a month, is it safe to say that this will not be going in and should be clsoed?

geky commented:
Truncate is a fundamental function for a POSIX filesystem. If it's not going in now, it should go in eventually. It's just seems to be a low priority item.

@cmonr, could be move this to a feature branch for now to get it out of the pr pool?

cmonr commented:
@geky Yeah, that works for me, unless anyone else opposes.

I can look into it tomorrow.

cmonr commented:
Branch `feature-posix` has been created. Feel free to close this PR and reopen against that branch.

geky commented:
@cmonr, turns out GitHub supports reassigning branches

cmonr commented:
@geky Take a look at the last Travis CI run. GitHub may work with it just bine, but from what I've observed, Travis tends to not like it.

We can see is that's still the case after a rebase ;)

cmonr commented:
Closing and re-opening to get Ttravis to recognize the updated base branch.

cmonr commented:
@kjbracey-arm Please take a look at the currently failing cloud client smoke test.

cmonr commented:
@davidsaada Still have concerns about this PR? @geky @ARMmbed/mbed-os-storage care to review?

0xc0170 commented:
> @davidsaada Still have concerns about this PR? @geky @ARMmbed/mbed-os-storage care to review?

This still needs to be reviewed.

davidsaada reviewed (approved):
Thanks for the fix. Looks good now.

geky commented:
Why don't we just merge this in 5.11?

cmonr commented:
/morph build

cmonr commented:
Hmmm.... Shot in the dark, but is it possible that the base branch needs to be rebased?

```
Scan: mbed-os-example-mesh-minimal
Scan: env
Traceback (most recent call last):
  File "/builds/ws/mbed-os-build-matrix/3445/DISCO_F469NI_GCC_ARM/sources/examples/mbed-os-example-mesh-minimal/mbed-os/tools/make.py", line 293, in <module>
    ignore=options.ignore
  File "/builds/ws/mbed-os-build-matrix/3445/DISCO_F469NI_GCC_ARM/sources/examples/mbed-os-example-mesh-minimal/mbed-os/tools/build_api.py", line 531, in build_project
    src_paths, toolchain, inc_dirs=inc_dirs)
  File "/builds/ws/mbed-os-build-matrix/3445/DISCO_F469NI_GCC_ARM/sources/examples/mbed-os-example-mesh-minimal/mbed-os/tools/resources/__init__.py", line 534, in scan_with_toolchain
    toolchain.config.load_resources(self)
  File "/builds/ws/mbed-os-build-matrix/3445/DISCO_F469NI_GCC_ARM/sources/examples/mbed-os-example-mesh-minimal/mbed-os/tools/config/__init__.py", line 1101, in load_resources
    self.validate_config()
  File "/builds/ws/mbed-os-build-matrix/3445/DISCO_F469NI_GCC_ARM/sources/examples/mbed-os-example-mesh-minimal/mbed-os/tools/config/__init__.py", line 1066, in validate_config
    raise error
UndefinedParameter: Attempt to override undefined parameter 'atmel-rf.provide-default' in 'application[*]'
```

NirSonnenschein commented:
also this:
[Error] main.cpp@26,0:  #35: #error directive: [NOT_SUPPORTED] CELLULAR_DEVICE must be defined
[DEBUG] Return: 1
[DEBUG] Output: "./features/cellular/TESTS/socket/udp/main.cpp", line 26: Error:  #35: #error directive: [NOT_SUPPORTED] CELLULAR_DEVICE must be defined
[DEBUG] Output: ./features/cellular/TESTS/socket/udp/main.cpp: 0 warnings, 1 error
Building project crc (DISCO_F469NI, ARM)
Scan: ARM
Scan: crc

kjbracey commented:
Fast-forwarded `feature-posix`, rebased `truncate`.

0xc0170 commented:
Travis failures look related please review

kjbracey commented:
Looks like automated merge failed - new methods now need some namespace specifiers.

cmonr commented:
/morph build

0xc0170 commented:
Test failures only for ARMCC and the test related to this changeset.

kjbracey commented:
Think that should fix it - test was saying "was our `seek` method the last thing to be called by `fseek(SEEK_END)`?". But I'd made it note `size` being called, and apparently ARM library calls `size` after `seek` for `SEEK_END`, not unreasonably, so test saw that. Moved check to `SEEK_SET` - presumably that should only make a `seek` method call.

0xc0170 commented:
/morph build

dannybenor commented on TESTS/mbed_platform/FileHandle/main.cpp:
```diff
@@ -420,10 +420,10 @@ void test_fseek_ftell()
     TestFile<FS>::resetFunctionCallHistory();
     fssek_ret = std::fseek(file, 0, SEEK_SET);
     TEST_ASSERT_EQUAL(0, fssek_ret);
+    TEST_ASSERT_TRUE(TestFile<FS>::functionCalled(TestFile<FS>::fnSeek));
```
It seems like an optimization, not calling seek if you are already there....

kjbracey replied:
Ah, could be. Already pushed another attempt which should clear it, covering that possibility too. It'll have to call seek for the `SEEK_END`.

cmonr commented:
/morph build

cmonr commented:
@kjbracey-arm When you get a chance, the test failures were quite consistent across devices.

kjbracey committed:
Rework fseek/ftell tests

ARM C library is really good at optimising out calls to underlying
seek. The only ones we were actually detecting in the empty file case
were the ones that the default FileHandle::size() made itself during
the SEEK_END case.

When we implement TestFile::size() directly, we will no longer see
a single seek call from the ARM C library in the empty file case, so
remove those tests.

Beef up the non-empty file case, adding checks that we are making
underlying read+write calls in the correct position, as a proxy for
direct checks for underlying seek being called.

kjbracey commented:
Okay, that was a pain. As @dannybenor suggested, the ARM C library is really good at optimising out underlying seek calls. In fact, the only ones the test was detecting in the ARM C empty file case were the ones that the default `FileHandle::size()` made itself. As soon as the `TestFile` was upgraded to have its own `size` for the truncate test there were no remaining seeks to detect in the empty file test. Reworked the test to cover everything in the non-empty file test.

I've run the modified tests locally on ARM C and GCC.

cmonr commented:
/morph build

kjbracey commented:
> Why don't we just merge this in 5.11?

Should I go ahead and make a merging PR for the feature branch? Could there be anything else that would be aiming for the feature branch in 5.11 timeframe?

cmonr commented:
Not that I'm aware of. @ARMmbed/mbed-os-storage ^^^

geky commented:
Probably better to aim for 5.12 at this point, but is there any reason not to move this into master?
